design/ceph-osd-removal.md

```diff
@@ -0,0 +1,113 @@
+# Removing OSDs from a Ceph cluster
+
+**Targeted for v1.X**
+
+## Background
+
+Removing OSDs is a common day-2 operation and users need to be able to perform
+this regularly.
+
+With the effort to add OSD replacement in the Ceph Dashboard, there is a demand
+for automating the deletion of individual OSDs.
+
+## Status
+
+As of Rook 1.1, there is some preliminary support for removing OSDs:
+
+### Removing Nodes
+
+Rook fully supports the removal of whole nodes from a Cluster, including all
+OSDs. This implementation does every call needed to completely remove all OSDs
+on that specific node from a Ceph cluster if that cluster no longer appears in
+the list of nodes in the CephCluster CR.
+
+There are concerns that this implementation might be dangerous, if a node is
+being removed inadvertently.
+
+### Removing individual OSDs
+
+Rook partially supports individual OSD-removal: The Rook operator monitors Ceph
+health and initiates the remove of OSDs that are, according to
+`ceph safe-to-destroy`, safe to be destroyed. Rook then removes the OSD
+Deployment automatically.
+
+Note that the device might stay referenced within the CephCluster CR.
+
+This feature is mainly aimed at automatically removing unresponsive OSDs.
+
+### Manual removal
+
+There is an existing Rook issue
+[#1827](https://github.com/rook/rook/issues/1827) containing a description of
+how to remove OSDs manually.
+
+## Design considerations
+
+Any future implementation within Rook needs to fulfill some requirements:
+
+1. To prevent inadvertent removal of OSDs, the act of removing OSDs needs to be
+   executed explicitly and manually.
+2. Omitting node or drive declarations within the CephCluster CR should not
+   physically remove OSDs.
+3. Except for physically removing data on disks, there should not be any manual
+   step required to allow a subsequent creation of a replacement OSD on the same
+   node.
+4. There should not be a perceived fundamental difference when removing whole
+   nodes versus removing single OSDs form a node.
+
+## Algorithm Outline
+
+Removing OSDs looks as follows when looking at the bare minimum of required
+steps (pseudocode):
+
+```python
+def ceph_crush_reweight(osd_id, weight): ...
+def ceph_pg_ls-by-osd(osd_id): ...
+def ceph_osd_out(osd_id): ...
+def ceph_osd_purge(osd_id): ...
+
+def remove_osd(osd_id, node, device_name):
+    ceph_crush_reweight(osd_id, 0.0)
+
+    while not len(ceph_pg-ls-by-osd(osd_id)):
+        time.sleep(10)
+
+    ceph_osd_out()
+
+    ceph_osd_purge()
+
+    assert ceph_safe-to-destroy()
+
+    delete_osd_deployment()
+
+    remove_osd_from_CephCluster_CR(node, device_name)
+```
+
+## Implementation Variant 1
+
+TODO: Is there a good way of implementing OSD-removal while accommodating the
```

Comment:
For the currently common, Kubernetes-focused Rook user, I think we probably just make sure we remove deployments for OSDs which are not present, but we shouldn't remove devices from CRUSH without manual user intervention via Ceph CLI/dashboard, possibly via `Variant 2` below.

Reply:
Can we make use of the CmdReporter to remove OSDs?

Reply:
That is the hope, yes. My thought was to use that to call `ceph-volume` to remove disks for removal/replacement.

Reply:
Q is: do we want to zap disks automatically?

Reply:
I think it would depend on the UX for replacing an OSD. If the UX is very careful about confirming that you really want to replace an OSD, maybe we should. What about this flow for replacing an OSD?
- The mgr module gets a request to replace an OSD. The request includes the OSD ID and the device name (e.g. `/dev/sdb`)
- The mgr module starts a CmdReporter job that does the following:
   - Checks if OSD ID was previously configured on the requested device. 
   - If the OSD ID was previously configured on the requested device:
       - Zap the disk
   - If the OSD ID was not previously configured on the requested device, check if the device is clean (no OSDs configured, no partitions, no filesystem)
      - If the device is not clean, abort the operation
   - Call ceph-volume to provision the OSD with the same ID on on the device that was just zapped
- Trigger an orchestration. The operator will find the new OSD has been provisioned and will start a new OSD deployment daemon for the replaced OSD.

Reply:
> I think it would depend on the UX for replacing an OSD. If the UX is very careful about confirming that you really want to replace an OSD, maybe we should. What about this flow for replacing an OSD?
> 
> * The mgr module gets a request to replace an OSD. The request includes the OSD ID and the device name (e.g. `/dev/sdb`)
> * The mgr module starts a CmdReporter job that does the following:
>   
>   * Checks if OSD ID was previously configured on the requested device.

We might no longer have that information in Ceph. In this case, I wouldn't do anything automatically

>   * If the OSD ID was previously configured on the requested device:
>     
>     * Zap the disk
>   * If the OSD ID was not previously configured on the requested device, check if the device is clean (no OSDs configured, no partitions, no filesystem)
>     
>     * If the device is not clean, abort the operation
>   * Call ceph-volume to provision the OSD with the same ID on on the device that was just zapped or otherwise already clean.

This needs to be done in a followup operation and needs to be added to the create-osd operation. As the user might need to physically touch the hardware, I don't see a way to automate this as one algorithm.

> * Trigger an orchestration. The operator will find the new OSD has been provisioned and will start a new OSD deployment daemon for the replaced OSD.

Reply:
Hm, we're really starting to get into a design document describing replacing osds here.

Reply:
> I think it would depend on the UX for replacing an OSD. If the UX is very careful about confirming that you really want to replace an OSD, maybe we should. What about this flow for replacing an OSD?
> 
> * The mgr module gets a request to replace an OSD. The request includes the OSD ID and the device name (e.g. `/dev/sdb`)
> * The mgr module starts a CmdReporter job that does the following:
>   
>   * Checks if OSD ID was previously configured on the requested device.
>   * If the OSD ID was previously configured on the requested device:
>     
>     * Zap the disk
>   * If the OSD ID was not previously configured on the requested device, check if the device is clean (no OSDs configured, no partitions, no filesystem)
>     
>     * If the device is not clean, abort the operation
>   * Call ceph-volume to provision the OSD with the same ID on on the device that was just zapped or otherwise already clean.

Ok, this doesn't work. Replacing OSD is always a two-stage process, as users need to manually replace hardware. Meaning, when creating new OSDs, we need to be able to make ceph-volume [reuse](https://docs.ceph.com/docs/master/mgr/orchestrator_modules/#ceph.deployment.drive_group.DriveGroupSpec.osd_id_claims) an existing osd id. 
> * Trigger an orchestration. The operator will find the new OSD has been provisioned and will start a new OSD deployment daemon for the replaced OSD.